completed authentication and profile data management

AfaqAhmed-XGrid opened the pull request:
Addresses #4

AfaqAhmed-XGrid committed:
completed authentication and profile data management

FahadJalal04 commented on config/passport.github.config.js:
```diff
@@ -0,0 +1,51 @@
+// Package imports
+const GitHubStrategy = require('passport-github2').Strategy;
```
These files should be in a separate directory called passport.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport.github.config.js:
```diff
@@ -0,0 +1,51 @@
+  );
+};
+
+module.exports = passportGithubConfig;
```
Add an extra line at the end of file.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport.google.config.js:
```diff
@@ -0,0 +1,52 @@
+
+// Constant Imports
+const constants = require('./config.constants');
+const User = require('../modules/auth/auth.model');
```
Not a constant.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport.google.config.js:
```diff
@@ -0,0 +1,52 @@
+    User.findOne({_id: id}).then((user) => cb(null, user));
+  });
+
+
```
Extra line.
Please configure eslint

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport.local.config.js:
```diff
@@ -0,0 +1,81 @@
+
+// Constant Imports
+const constants = require('./config.constants');
+const User = require('../modules/auth/auth.model');
```
Not a constant import

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport.local.config.js:
```diff
@@ -0,0 +1,81 @@
+        try {
+          const user = await User.findOne({email});
+          if (!user) {
+            console.log('first');
```
Use winston log for logging

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport.local.config.js:
```diff
@@ -0,0 +1,81 @@
+          const user = await User.findOne({email});
+          if (!user) {
+            console.log('first');
+            return done(null, false, {message: constants.WRONG_EMAIL_MSG});
```
**WRONG_EMAIL_MSG** use camelCase

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport.local.config.js:
```diff
@@ -0,0 +1,81 @@
+      new localStrategy(
+          {passReqToCallback: true, usernameField: 'email'},
+          async function(req, email, password, done) {
+            try {
```
use await to instead of try-catch. 
For ref: https://www.npmjs.com/package/await-to-js

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on index.js:
```diff
@@ -1,21 +1,60 @@
+const passport = require('passport');
+const cookieParser = require('cookie-parser');
+const bodyParser = require('body-parser');
+const cors = require('cors');
```
Define custom CORS

AfaqAhmed-XGrid replied:
Done

FahadJalal04 replied:
Remove this import statement if custom cors are defined.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on constants/constants.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  RUNNING_APP_MSG: 'App is successfully running on',
+  FAILED_RUNNING_APP_MSG: 'Error occurred! Server cannot start ',
```
Suggestion: refactor it like:
{
  responseMessages: {
    authenticateUser: {
      success: '...',
      failure: '...'
    },
  }
}

AfaqAhmed-XGrid replied:
Done

AfaqAhmed-XGrid committed:
resolved all issues in commit 1

used logger, changed to camelcase in constants files, defined custom cors and updated the folder structure

FahadJalal04 commented on config/constants.js:
```diff
@@ -0,0 +1,53 @@
+      },
+      findingUser: {
+        error: 'Error while finding the user if its already existed in db (github)',
+        found: 'User is existed in db (github)',
```
```suggestion
        error:
        found:
```
There should be only two keys success and failure. Please update this file accordingly.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 replied:
Not addressed

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/constants.js:
```diff
@@ -0,0 +1,53 @@
+      savingNewUser: {
+        error: 'Error in saving user through (github strategy)',
+        success: 'New user is saved in db through (github strategy)',
+        failure: 'New user could not be saved in db through (github strategy)',
```
What is the difference between error and failure? Same comment as line 13

AfaqAhmed-XGrid replied:
Resolved the issue

FahadJalal04 commented on config/db.connect.js:
```diff
@@ -0,0 +1,21 @@
+const dbConnect = async () => {
+  const [err, success] = await to(mongoose.connect(process.env.MONGODB_URL, {
+    useNewUrlParser: true,
+    useUnifiedTopology: true,
```
```suggestion
    useUnifiedTopology: true,
```
Please explain why are you using this?

AfaqAhmed-XGrid replied:
it is a type of mechanism that mongoose uses to establish connection with mongodb servers. In this mechanism/engine, the pool of connections is created that can be shared across all mongodb servers. it improves the mongodb performance.

FahadJalal04 replied:
Ack

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/logger/logger.js:
```diff
@@ -0,0 +1,30 @@
+  transports: [
+    new transports.File({filename: 'error.log', level: 'error'}),
+    new transports.File({filename: 'quick-start-combined.log'}),
+  ],
```
```suggestion
defaultMeta: {service: 'user-service'},
  transports: [
    new transports.File({filename: 'error.log', level: 'error'}),
    new transports.File({filename: 'quick-start-combined.log'}),
  ],
```
Explain this.

AfaqAhmed-XGrid replied:
it is the metadata that would be included in all the logger messages. I added it just to see how it works

FahadJalal04 replied:
Ack

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport/passport.github.config.js:
```diff
@@ -0,0 +1,75 @@
+const to = require('await-to-js').default;
+
+// Constant imports
+const globalConstants = require('../../constants/constants');
```
```suggestion
const globalConstants = require('../../constants/constants');
```
Rename it to constants

```suggestion
const configConstants = require('../constants');
```
Rename it to gitHubConstants

Please use meaningful names

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport/passport.github.config.js:
```diff
@@ -0,0 +1,75 @@
+              return done(null, false, {message: globalConstants.responseMessages.logInUser.failure});
+            }
+
+            if (existingUser) {
```
```suggestion
            if (existingUser) 
```
Please add a comment before it explaining why are you doing this.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport/passport.github.config.js:
```diff
@@ -0,0 +1,75 @@
+              githubId: profile?.id,
+              displayName: profile?.displayName,
+              profilePicture: profile?.photos[0]?.value,
+              email: profile?.id + '@email.com',
```
```suggestion
              '@email.com'
```
Is this intentional?

AfaqAhmed-XGrid replied:
Yes! Actually, I had this idea even if the user is signing in through google or github, he can change his profile data, add email, set password and can then use the account as the local one. So, when the user sign in through google/github, what he would see as email in its profile data initially?? I cannot set a common email to all because in schema i ahve set the email to be unique. that's why, I am doing this, saving the profile in the email format in place of actual email. then the user can update the email if he wants to.

FahadJalal04 replied:
If someone is logging in from Google he/she can't change password/email locally. Please change this implementation and use the information Google is returning. If a user changes the Google profile name this change should be reflected here.
This implementation is not correct if we are authentication users via Google and Github.

AfaqAhmed-XGrid replied:
Done

AfaqAhmed-XGrid committed:
updated the constants

arxlanXgrid40 commented on config/db.connect.js:
```diff
@@ -0,0 +1,19 @@
+// Constant imports
+const constants = require('./config.constants');
+
+const dbConnect = async () => {
```
always add functional comments

AfaqAhmed-XGrid replied:
done

FahadJalal04 replied:
This is not a function-level comment.

AfaqAhmed-XGrid replied:
DOne

arxlanXgrid40 commented on config/passport.local.config.js:
```diff
@@ -0,0 +1,81 @@
+        try {
+          const user = await User.findOne({email});
+          if (!user) {
+            console.log('first');
```
remove this

AfaqAhmed-XGrid replied:
done

FahadJalal04 commented on config/logger/logger.js:
```diff
@@ -0,0 +1,30 @@
+const {createLogger, format, transports} = require('winston');
+
+const logger = createLogger({
+  level: 'info',
```
Why level is hardcoded? It shouldn't be hardcoded this should be modified according to the situation (Error, debug, info).

AfaqAhmed-XGrid replied:
removed it!

FahadJalal04 commented on config/logger/logger.js:
```diff
@@ -0,0 +1,30 @@
+  level: 'info',
+  format: format.combine(
+      format.timestamp({
+        format: 'YYYY-MM-DD HH:mm:ss',
```
```suggestion
        format: 'YYYY-MM-DD HH:mm:ss',
```
Format should be referenced from constants

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport/passport.google.config.js:
```diff
@@ -0,0 +1,76 @@
+            const [newUserSavedError, newUserSaved] = await to(newUser.save());
+
+            if (newUserSavedError) {
+              logger.error(googleConstants.labels.googlePassportStrategy.savingNewUser.failure, newUserSavedError);
```
You are only logging errors. Please add a log for what is happening for example API server is about to add a new user in DB there should be a log statement indicating that the API server is adding a new user and after a successful request add a log statement telling the user is successfully added.

AfaqAhmed-XGrid replied:
Done

AfaqAhmed-XGrid committed:
added some logs and changed some logger properties

FahadJalal04 commented on .env.example:
```diff
@@ -0,0 +1,18 @@
+GOOGLE_CLIENT_SECRET = GOOGLE_CLIENT_SECRET
+GOOGLE_CLIENT_ID = GOOGLE_CLIENT_ID
+GOOGLE_CALLBACK_URL = GOOGLE_CALLBACK_URL
+GITHUB_CLIENT_ID = GITHUB_CLIENT_ID
+GITHUB_CLIENT_SECRET = GITHUB_CLIENT_SECRET
+GITHUB_CALLBACK_URL = GITHUB_CALLBACK_URL
+MONGODB_URL = MONGODB_URL
+SESSION_SECRET_KEY = SESSION_SECRET_KEY
+JWT_SECRET = JWT_SECRET
+PORT = PORT
+CLIENT_URL = CLIENT_URL
+SESSION_CODE = SESSION_CODE
+EMAIL_HOST = EMAIL_HOST
+EMAIL_PORT = EMAIL_PORT
+EMAIL_USER = EMAIL_USER
+EMAIL_PASS = EMAIL_PASS
+EMAIL_FROM = EMAIL_FROM
+NODE_ENV = development
```
This file should be renamed as `.env`

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/db.connect.js:
```diff
@@ -0,0 +1,24 @@
+
+// Connecting mongodb server using mongoose
+const dbConnect = async () => {
+  console.log('hello: ', process.env.MONGODB_URL);
```
use Winston log

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport/passport.google.config.js:
```diff
@@ -0,0 +1,75 @@
+// Package imports
+const GoogleStrategy = require('passport-google-oauth20').Strategy;
```
Not a package import

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport/passport.google.config.js:
```diff
@@ -0,0 +1,75 @@
+// Logger import
+const logger = require('../logger/logger');
+
+const passportGoogleConfig = (passport) => {
```
Function level comment missing. Please add function level comments to all your functions.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on config/passport/passport.local.config.js:
```diff
@@ -0,0 +1,91 @@
+                });
+              } else {
+                logger.error(localConstants.labels.localPassportStrategy.signup.savingNewUser.failure, newUserSavingError);
+                return done(null, false, {message: constants.responseMessages.signUpUser.failure});
```
add space in brackets rule in the eslint config file.
For ref: https://eslint.org/docs/latest/rules/space-in-brackets

AfaqAhmed-XGrid replied:
'object-curly-spacing': [ 'error', 'always' ],
 'space-in-parens': [ 'error', 'always' ],
 'array-bracket-spacing': [ 'error', 'always' ],
 
 Added these rules

FahadJalal04 commented on config/passport/passport.local.config.js:
```diff
@@ -0,0 +1,91 @@
+  });
+};
+
+module.exports = passportLocalConfigure;
```
Use eol-last to enforce at least one line at the end of the file.
For ref: https://eslint.org/docs/latest/rules/eol-last

and use no-multiple-empty-lines rule to prevent adding multiple empty lines.
For ref: https://eslint.org/docs/latest/rules/no-multiple-empty-lines

AfaqAhmed-XGrid replied:
added

FahadJalal04 commented on error.log:
Why this file is added to the commit?

AfaqAhmed-XGrid replied:
removed

FahadJalal04 commented on index.js:
```diff
@@ -1,21 +1,66 @@
+passportGithubConfig(passport);
+
+// Configuring Routes
 app.use('/api/practice/', practiceRouter);
+app.use('/api/auth/', authRouter);
 
```
These routes should be defined in the routes file as discussed earlier.

AfaqAhmed-XGrid replied:
these are not the routes being defined here. I ma just integrating created routes in app through app.use()

FahadJalal04 commented on modules/auth/auth.constants.js:
```diff
@@ -0,0 +1,130 @@
+      resetPassword: {
+        success: 'Your password has been reset successfully',
+        failure: 'Failed to reset password. Please try again!',
+        tokenExpiredMessage: 'Your token has expired. Please try again',
```
```suggestion
        tokenExpiredMessage: 'Your session has expired. Please try again',
```

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on modules/auth/auth.constants.js:
```diff
@@ -0,0 +1,130 @@
+      },
+      resetPasswordEmail: {
+        success: 'Password reset mail is sent to your account',
+        failure: 'Password reset mail could not be sent! Check your network connectivity!',
```
How do you know it's a connectivity issue?
The message should be generic.

AfaqAhmed-XGrid replied:
done

FahadJalal04 commented on modules/auth/auth.constants.js:
```diff
@@ -0,0 +1,130 @@
+      resetPasswordEmail: {
+        success: 'Password reset mail is sent to your account',
+        failure: 'Password reset mail could not be sent! Check your network connectivity!',
+        emailText: `Please click on the link to reset your password. The link is active for 5 minuts only!`,
+        emailSubject: 'Reset you password!',
```
Email subject and body should not be added in response messages.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on modules/auth/auth.constants.js:
```diff
@@ -0,0 +1,130 @@
+      },
+    },
+    userProfileData: {
+      success: 'Your data has been fetched successfully',
```
Which data has been fetched successfully?
Messages should be meaningful.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on modules/auth/auth.constants.js:
```diff
@@ -0,0 +1,130 @@
+      failure: 'Failed to fetch data. Please try again!',
+    },
+    updateUserProfile: {
+      success: 'Your data has been updated successfully',
```
same

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on modules/auth/auth.constants.js:
```diff
@@ -0,0 +1,130 @@
+    },
+  },
+  labels: {
+    session: 'session: ',
+    sessiodId: 'sessionId: ',
+    user: 'user: ',
+    error: 'error: ',
+    info: 'info: ',
+    middleware: 'middleware: ',
```
Where are you using these labels?

AfaqAhmed-XGrid replied:
Issue is resolved

FahadJalal04 commented on modules/auth/auth.constants.js:
```diff
@@ -0,0 +1,130 @@
+    error: 'error: ',
+    info: 'info: ',
+    middleware: 'middleware: ',
+    loginCtrl: {
+      failure: 'user is not found in request. login() problem. (loginCtrl)',
+      success: 'User is loggedin',
+    },
+    signupCtrl: {
+      success: 'User is signed ip',
+      failure: 'user is not found in request. login() problem. (signupCtrl)',
+    },
+    logoutCtrl: {
+      failure: 'Error while logging out the user (logoutCtrl)',
+      success: 'User is loggedout',
+    },
+    changePasswordCtrl: {
+      noUserInReq: 'user is not found in request. login() problem. (changePasswrodCtrl)',
+      findingUser: {
+        failure: 'Error while finding user (changePasswordCtrl)',
+      },
+      savingNewPassword: {
+        failure: 'Error while saving new password (changePasswordCtrl)',
+        success: 'saved new password in db',
+        savingMsg: 'Saving new password in db',
+      },
+    },
+    forgotPasswordCtrl: {
+      findingUser: {
+        failure: 'Error while finding user (forgotPasswordCtrl)',
+      },
```
Why response messages are added in the labels field?

AfaqAhmed-XGrid replied:
Issue is resolved

FahadJalal04 commented on modules/auth/auth.controller.js:
```diff
@@ -0,0 +1,296 @@
+ * @param {Object} res
+ * @return {Object}
+ */
+const loginCtrl = async (req, res) => {
```
```suggestion
const login = async (req, res) => {
```
The name should be proper. No need to add ctrl in the name as it is already in the controller file.

FahadJalal04 replied:
Same for all functions in the controller files.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on modules/auth/auth.controller.js:
```diff
@@ -0,0 +1,296 @@
+ * @return {Object}
+*/
+const signupCtrl = async (req, res, next) => {
+  if (req.user) {
+    logger.error(authConstants.labels.signupCtrl.success, req);
+    return res.status(200).json({success: 1, message: globalConstants.responseMessages.signUpUser.success, data: req.user});
+  } else {
+    logger.error(authConstants.labels.signupCtrl.failure, req);
+    return res.status(200).json({success: 0, message: globalConstants.responseMessages.signUpUser.failure, data: null});
+  }
```
Please explain what are you doing in the controller function. `loginCtrl` and `signupCtrl` both controller functions are same with different names.

AfaqAhmed-XGrid replied:
The response is different in both controllers in terms of the message we are sending

FahadJalal04 commented on modules/auth/auth.controller.js:
```diff
@@ -0,0 +1,296 @@
+  if (passwordSaved) {
+    logger.info(authConstants.labels.changePasswordCtrl.savingNewPassword.success, req.user);
+    return res.status(200).json({success: 1, message: authConstants.responseMessages.changePassword.success, data: user});
+  } else {
```
```suggestion
```
No need to write `else` as you already added the return statement in `if`

AfaqAhmed-XGrid replied:
done

AfaqAhmed-XGrid committed:
Eslint new rules, improved responsed msgs, changed naming conventions, changed logging pattern, added env example file

Changed response messages to more meaningful ones.
Changed logging pattern
Added some eslint rules
Change naming convention for controllers and middlewares

AfaqAhmed-XGrid committed:
Revert "Eslint new rules, improved responsed msgs, changed naming conventions, changed logging pattern, added env example file"

This reverts commit d81fbd7ad3cc146d85b439bb07b2eb4cd01e456b.

AfaqAhmed-XGrid committed:
Revert "Revert "Eslint new rules, improved responsed msgs, changed naming conventions, changed logging pattern, added env example file""

This reverts commit 8f61be62acbfbf4d2c052f9b5c36f24e5e976847.

FahadJalal04 commented on config/passport/passport.github.config.js:
```diff
@@ -0,0 +1,78 @@
+
+/**
+ * Defining the passport github strategy
+ * @param {any} passport
```
Query: Why passport type is set to any?

AfaqAhmed-XGrid replied:
Changed to actual type of passport

FahadJalal04 commented on .eslintrc.js:
```diff
@@ -15,7 +15,12 @@ module.exports = {
     'linebreak-style': 'off',
+    'object-curly-spacing': [ 'error', 'always' ],
+    'space-in-parens': [ 'error', 'always' ],
+    'array-bracket-spacing': [ 'error', 'always' ],
```
Remove this rule. No need to add this. Recommended to add this rule`object-curly-spacing`  only.

AfaqAhmed-XGrid replied:
Done

FahadJalal04 commented on modules/auth/auth.route.js:
```diff
@@ -0,0 +1,68 @@
+const router = express.Router();
+
+// Creating Routes
+router.post(
+    '/login',
+    authMiddleware.login,
+    authController.login,
+);
+router.post(
+    '/signup',
+    authMiddleware.signup,
+    authController.signup,
+);
+router.get(
+    '/logout',
+    authMiddleware.logout,
+    authController.logout,
+);
+router.put(
+    '/change-password',
+    authMiddleware.isAuthenticated,
+    authController.changePassword,
+);
```
Add a comment before each route explaining the functionality of the endpoint.

AfaqAhmed-XGrid replied:
Done

AfaqAhmed-XGrid committed:
remove square and paren bracket rules, added comments for all auth endpoints, added proper param types in function level comments

FahadJalal04 commented on config/logger/logger.js:
```diff
@@ -0,0 +1,30 @@
+// Package imports
+const { createLogger, format, transports } = require('winston' );
```
```suggestion
const { createLogger, format, transports } = require('winston');
```

FahadJalal04 commented on config/logger/logger.js:
```diff
@@ -0,0 +1,30 @@
+// Package imports
+const { createLogger, format, transports } = require('winston' );
+const constants = require('../constants' );
```
```suggestion
const constants = require('../constants');
```

FahadJalal04 commented on config/logger/logger.js:
```diff
@@ -0,0 +1,30 @@
+  format: format.combine(
+      format.timestamp({
+        format: constants.logger.dateFormat,
+      } ),
```
```suggestion
      }),
```

AfaqAhmed-XGrid replied:
Resovled

AfaqAhmed-XGrid committed:
resolved paren spacing issue

AfaqAhmed-XGrid committed:
Update auth.helper.js
